Android/app/src/main/java/app/intra/socks/UdpOverrideSocksHandler.java

```diff
@@ -0,0 +1,95 @@
+package app.intra.socks;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+import sockslib.common.Socks5DatagramPacketHandler;
+import sockslib.common.SocksException;
+import sockslib.server.Session;
+import sockslib.server.Socks5Handler;
+import sockslib.server.UDPRelayServer;
+import sockslib.server.msg.CommandMessage;
+import sockslib.server.msg.CommandResponseMessage;
+import sockslib.server.msg.ServerReply;
+
+/**
+ * This SocksHandler acts as a normal Socks5Handler, except that UDP packets to |fakeDns| are
+ * actually redirected to |trueDns|.
+ */
+public class UdpOverrideSocksHandler extends Socks5Handler {
+  private InetSocketAddress fakeDns = null;
+  private InetSocketAddress trueDns = null;
+
+  // This class is instantiated by a call to newInstance() in
+  // BasicSocksProxyServer.createSocksHandler() (i.e. via the default constructor).
+  // We can't provide it with the DNS information in the constructor, so we need a new method for
+  // the purpose.
+
+  /**
+   * @param fakeDns Where the system thinks the DNS server is.  Typically has port 53.
+   * @param trueDns Where the DNS server actually is.  Typically a high-numbered port on localhost.
+   */
+  void setDns(InetSocketAddress fakeDns, InetSocketAddress trueDns) {
+    this.fakeDns = fakeDns;
+    this.trueDns = trueDns;
+  }
+
+  @Override
+  public void doUDPAssociate(Session session, CommandMessage commandMessage) throws IOException {
+    // Construct a UDP relay server, just like upstream.
+    UDPRelayServer udpRelayServer =
+        new UDPRelayServer(((InetSocketAddress) session.getClientAddress()).getAddress(),
+            commandMessage.getPort());
+
+    // Set buffer size to 5 KB.  (The default is 5 MB, which causes OOM crashes.)
+    final int BUFFER_SIZE = 5 * 1024;
```

Comment:
Could you make these final variables static constants?

Reply:
Done